server/sonar-db-dao/src/main/java/org/sonar/db/permission/PermissionQuery.java

```diff
@@ -0,0 +1,178 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.db.permission;
+
+import java.util.Locale;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+import org.sonar.db.WildcardPosition;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Preconditions.checkArgument;
+import static org.apache.commons.lang.StringUtils.defaultIfBlank;
+import static org.sonar.api.utils.Paging.offset;
+import static org.sonar.db.MoreDatabaseUtils.buildLikeValue;
+
+/**
+ * Query used to get users and groups permissions
+ */
+@Immutable
+public class PermissionQuery {
+  public static final int RESULTS_MAX_SIZE = 100;
+  public static final int SEARCH_QUERY_MIN_LENGTH = 3;
+  public static final int DEFAULT_PAGE_SIZE = 20;
+  public static final int DEFAULT_PAGE_INDEX = 1;
+
+  // filter: return only the users or groups who have this permission
+  private final String permission;
+  // filter on project, else filter global permissions
+  private final String componentUuid;
+  private final String template;
+
+  // filter on login, email or name of users or groups
+  private final String searchQuery;
+  private final String searchQueryToSql;
+  private final String searchQueryToSqlLowercase;
+
+  // filter users or groups who have at least one permission. It does make
+  // sense when the filter "permission" is set.
+  private final boolean withAtLeastOnePermission;
+
+  private final int pageSize;
+  private final int pageOffset;
+
+  private PermissionQuery(Builder builder) {
+    this.permission = builder.permission;
+    this.withAtLeastOnePermission = builder.withAtLeastOnePermission;
+    this.componentUuid = builder.componentUuid;
+    this.template = builder.template;
+    this.searchQuery = builder.searchQuery;
+    this.searchQueryToSql = builder.searchQuery == null ? null : buildLikeValue(builder.searchQuery, WildcardPosition.BEFORE_AND_AFTER);
+    this.searchQueryToSqlLowercase = builder.searchQuery == null ? null : searchQueryToSql.toLowerCase(Locale.ENGLISH);
+    this.pageSize = builder.pageSize;
+    this.pageOffset = offset(builder.pageIndex, builder.pageSize);
+  }
+
+  @CheckForNull
+  public String getPermission() {
+    return permission;
+  }
+
+  public boolean withAtLeastOnePermission() {
+    return withAtLeastOnePermission;
+  }
+
+  // TODO remove it, it should not be in the query, but set as a separate parameter
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)